test/integration/ads_integration_test.cc

```diff
@@ -992,4 +992,202 @@ TEST_P(AdsIntegrationTestWithRtdsAndSecondaryClusters, Basic) {
+                                   envoy::config::core::v3::ApiVersion::V3)) {
+      return TestUtility::anyConvert<envoy::config::route::v3::RouteConfiguration>(resource).name();
+    }
+    if (resource.type_url() == Config::getTypeUrl<envoy::config::cluster::v3::Cluster>(
+                                   envoy::config::core::v3::ApiVersion::V3)) {
+      return TestUtility::anyConvert<envoy::config::cluster::v3::Cluster>(resource).name();
+    }
+    if (resource.type_url() ==
+        Config::getTypeUrl<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+            envoy::config::core::v3::ApiVersion::V3)) {
+      return TestUtility::anyConvert<envoy::config::endpoint::v3::ClusterLoadAssignment>(resource)
+          .cluster_name();
+    }
+    if (resource.type_url() == Config::getTypeUrl<envoy::config::route::v3::VirtualHost>(
+                                   envoy::config::core::v3::ApiVersion::V3)) {
+      return TestUtility::anyConvert<envoy::config::route::v3::VirtualHost>(resource).name();
+    }
+    if (resource.type_url() == Config::getTypeUrl<envoy::service::runtime::v3::Runtime>(
+                                   envoy::config::core::v3::ApiVersion::V3)) {
+      return TestUtility::anyConvert<envoy::service::runtime::v3::Runtime>(resource).name();
+    }
+    throw EnvoyException(
+        absl::StrCat("xdsResourceName does not know about type URL ", resource.type_url()));
+  }
+
+  template <class T>
+  void sendDiscoveryResponse(const std::string& type_url, const std::vector<T>& state_of_the_world,
+                             const std::vector<T>& added_or_updated,
+                             const std::vector<std::string>& removed, const std::string& version) {
+    if (sotw_or_delta_ == Grpc::SotwOrDelta::Sotw) {
+      sendSotwDiscoveryResponse(type_url, state_of_the_world, version);
+    } else {
+      sendDeltaDiscoveryResponse(type_url, added_or_updated, removed, version);
+    }
+  }
+
+  template <class T>
+  void sendSotwDiscoveryResponse(const std::string& type_url, const std::vector<T>& messages,
+                                 const std::string& version) {
+    API_NO_BOOST(envoy::service::discovery::v3::DiscoveryResponse) discovery_response;
+    discovery_response.set_version_info(version);
+    discovery_response.set_type_url(type_url);
+    for (const auto& message : messages) {
+      discovery_response.add_resources()->PackFrom(message);
+    }
+    static int next_nonce_counter = 0;
+    discovery_response.set_nonce(absl::StrCat("nonce", next_nonce_counter++));
+    xds_stream_->sendGrpcMessage(discovery_response);
+  }
+
+  template <class T>
+  void
+  sendDeltaDiscoveryResponse(const std::string& type_url, const std::vector<T>& added_or_updated,
+                             const std::vector<std::string>& removed, const std::string& version) {
+    sendDeltaDiscoveryResponse(type_url, added_or_updated, removed, version, xds_stream_);
+  }
+
+  template <class T>
```

Comment:
Why are these templated? Aren't they all v3? Do we need these wrappers?

Reply:
I'm still a bit confused; these belong to `AdsClusterV3Test`, which is only doing v3 AFAICT. So, why do they need templating?

Reply:
these were copied from their counterpartes on `BaseIntegrationTest` but i'm not really sure why they all need templating at all, they could just receive generic `Protobuf::Message`.

Reply:
The templating is not for the api version but for the resource type this methods receive clusters, cluster load assignements, routes, etc.

Reply:
I'm still wondering if we really need these. There is a transport/resource version distinction in xDS, here we only care about v3 xDS resource API version, there should be no need to have specialized methods for the v3 API transport to validate the pause/resume behavior.

Reply:
the reason i added these is to override the default resource version downgrading [here](https://github.com/envoyproxy/envoy/blob/master/test/integration/integration.h#L327) and [here](https://github.com/envoyproxy/envoy/blob/master/test/integration/integration.h#L292)

Reply:
i could probably make those two methods virtual and override only those, would that be ok?

Reply:
Sure, or even just parameterize the downgrade in the function args.